### PR TITLE
FIX: bounded parameter in cdfchn.f gives bad results

### LINKED ISSUE #2158: special.chndtrix  (ncx2.ppf) gives wrong results (Trac #1633)
_Original ticket http://projects.scipy.org/scipy/ticket/1633 on 2012-03-26 by trac user pwei, assigned to @pv._

ncx2.ppf(1.-1e-6,2,1) --> 35.4[[BR]]

ncx2.ppf(1.-1e-6,2,2) --> 1e+100 <======[[BR]]

ncx2.ppf(1.-1e-6,2,3) --> 45.98[[BR]]

ncx2.ppf(1.-1e-6,2,4) --> 1e+100 <======[[BR]]

ncx2.ppf(1.-1e-6,2,5) --> 52.67

### LINKED ISSUE #8665: `scipy.ncx2.sf` should be monotone decreasing
I'm getting unexpected behavior from `scipy.ncx2.sf`.  The survival function should eventually go to zero for large enough inputs, but instead I'm seeing a kind of "foot", i.e., it is bottoming out at a positive value.  The following interactive I/O demonstrates the issue.

### Reproducing code **example:**

(Pdb) v_T_norm[::50]
array([  4.99515382e+00,   1.07617327e+01,   2.31854502e+01,
         4.99515382e+01,   1.07617327e+02,   2.31854502e+02,
         4.99515382e+02,   1.07617327e+03,   2.31854502e+03,
         4.99515382e+03,   1.07617327e+04,   2.31854502e+04,
         4.99515382e+04])
(Pdb) nu
20.0
(Pdb) lam
499.51538166556196
(Pdb) ncx2.sf(v_T_norm[::50], df=nu, nc=lam)
array([  1.00000000e+00,   1.00000000e+00,   1.00000000e+00,
         1.00000000e+00,   1.00000000e+00,   1.00000000e+00,
         6.64666099e-01,   4.16325934e-05,   4.16325934e-05,
         4.16325934e-05,   4.16325934e-05,   4.16325934e-05,
         4.16325934e-05])

### LINKED ISSUE #9836: scipy.stats.rice gives incorrect results when s is very low compared to R
According to the [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rice.html):  
image

As you will see below, this assumption does not seem to hold when `R` is much bigger than `s`.  

I wrote some functions to generate a `(U, V)` vector and `(U + u, V + v)` random deviates. In each case I compute the percent point function with `rice.ppf` and I compare the results with the empirical distribution from `n = 10000` random deviates.  

### Simulation code
This is for reference, you can skip it first.  
```python
from itertools import product

from numpy import sqrt, mean, split, linspace, quantile, asarray, absolute, isnan, column_stack, around
from numpy.random import normal, choice
from scipy.stats import rice

from plotly.offline import init_notebook_mode, iplot
import plotly.graph_objs as go

init_notebook_mode()


def plot_input(U, V, u, v):
    sample = choice(u.size, min(u.size, 1000))
    traces = [go.Scatter(x=U + u[sample], y=V + v[sample], mode='markers', marker=dict(color='#377EB8', size=4, opacity=1/3))]
    layout = go.Layout(
        title='Initial vector and random deviates', width=700, height=400, yaxis=dict(scaleanchor="x"),
        shapes = [dict(type='line', x0=0, x1=U, y0=0, y1=V, line = dict(color='#4DAF4A', width=2))], 
        margin = go.layout.Margin(l=50, r=50, b=50, t=50)
    )
    return go.Figure(data=traces, layout=layout)


def plot_distribution(R, probabilities, empirical_quantiles, rice_quantiles):
    traces = [
        go.Scatter(x=empirical_quantiles, y=probabilities, marker=dict(color='#377EB8'), name='empirical'), 
        go.Scatter(x=rice_quantiles, y=probabilities, marker=dict(color='#4DAF4A'), name='rice')
    ]
    layout = go.Layout(
        title='Cumulative distribution functions', width=700, height=400, 
        shapes=[dict(type='line', x0=R, x1=R, y0=0, y1=1, line = dict(color='#E41A1C', width=1))], 
        margin = go.layout.Margin(l=50, r=50, b=50, t=50)
    )
    return go.Figure(data=traces, layout=layout)


def simulation(R, s, n=10000, U=0, plots=True):
    # Compute V so that sqrt(U**2 + V**2) == R
    V = sqrt(R**2 - U**2)
    if isnan(V):
        raise ValueError('invalid inputs')
    # Draw random samples from a normal (Gaussian) distribution with standard deviation s
    u, v = split(normal(scale=s, size=n * 2), 2)
    # compute norms of the vectors
    distances = sqrt((U + u)**2 + (V + v)**2)
    # get quantiles from the empirical and ricean distributions
    probabilities = linspace(0, 1, 101)[1:-1]
    empirical_quantiles = quantile(distances, probabilities)
    rice_quantiles = rice.ppf(probabilities, R / s, scale=s)
    output = dict(
        R=R, s=s, n=n, U=U, V=V, probabilities=probabilities, empirical_quantiles=empirical_quantiles, 
        rice_quantiles=rice_quantiles, relative_error=(rice_quantiles - empirical_quantiles) / empirical_quantiles
    )
    if plots:
        output['plots'] = {
            'input': plot_input(U, V, u, v),
            'distribution': plot_distribution(R, probabilities, empirical_quantiles, rice_quantiles)
        }
    return output    
```

### Examples
With `R = 50` and `s = 4` the result is visually OK:  
```python
sim = simulation(50, 4, U=47.5)
iplot(sim['plots']['input'])
iplot(sim['plots']['distribution'])
```
image

But for `R = 500` and `s = 4` the quantiles are off by about `100` units:  
```python
sim = simulation(500, 4, U=475)
iplot(sim['plots']['input'])
iplot(sim['plots']['distribution'])
```
image

I'm pretty sure I am missing some point here (probably something with the `loc` parameter), but in any case from the documentation above I would except from `rice.ppf` to produce a curve centered around `500`. Why is it not the case ?  

### More simulations
I tried to understand my issue by making more simulations. I computed the mean relative error for each simulation and plotted the results. I also made a plot of the relative error by the ratio of `R/s`.  
```python
distances = linspace(0, 500, 11)
deviations = linspace(.25, 5, 10)
simulations = []
for R, s in product(distances, deviations):
    simulations.append((R, s, mean(simulation(R, s, plots=False)['relative_error'])))
simulations = asarray(simulations)
simulations = simulations[(simulations[:, 0] / simulations[:, 1]).argsort()]

traces = [go.Scatter(
    x=simulations[:, 0], y=simulations[:, 1], mode='markers',
    marker=dict(
        color=simulations[:, 2], showscale=True, opacity=3/4,
        colorscale='Viridis', size=20
    )
)]
layout = go.Layout(
    title='Mean relative error', xaxis=dict(title='distance'), yaxis=dict(title='sigma'),
    width=900, height=500, margin = go.layout.Margin(l=50, r=50, b=50, t=50)
)
iplot(go.Figure(data=traces, layout=layout))

traces = [go.Scatter(x=simulations[:, 0] / simulations[:, 1], y=simulations[:, 2])]
layout = go.Layout(
    title='Mean relative error in function of the length over variance ratio', width=900, height=500, 
    xaxis=dict(title='R/s'), yaxis=dict(title='mean relative error', autorange='reversed')
)
iplot(go.Figure(data=traces, layout=layout))
```
image

It looks like the results suddenly go off if the ratio goes above 100!  

### Comparison with R
For information, I compared the results produced by `rice.ppf` and `rice.pdf` to, respectively, the `qrice` and `price` functions in R's `VGAM` package:
```python
from pandas import DataFrame

probabilities = linspace(0, 1, 11)[1:-1]
scipy_rice_quantiles = around(rice.ppf(probabilities, 500/4, scale=4), 6)
# qrice function in VGAM R package
# `qrice(seq(0.1, .9, by = .1), vee = 500, sigma = 4)`
vgam_qrice_quantiles = [494.889876, 496.649569, 497.918432, 499.002628, 500.016, 501.029372, 502.113568, 503.382431, 505.142125]
DataFrame({'probabilities': probabilities, 'scipy': scipy_rice_quantiles, 'VGAM': vgam_qrice_quantiles})
```
image

```python
quantiles = linspace(480, 520, 11)
scipy_rice_probabilities = around(rice.cdf(quantiles, 500/4, scale=4), 6)
# price function in VGAM R package
# `price(seq(480, 520, length.out = 11), vee = 500, sigma = 4)`
vgam_price_probabilities = [0.000000, 0.000031, 0.001332, 0.022533, 0.157685, 0.498404, 0.840379, 0.977035, 0.998632, 0.999968, 1.000000]
DataFrame({'quantiles': quantiles, 'scipy': scipy_rice_probabilities, 'VGAM': vgam_price_probabilities})
```
image

### Scipy/Numpy/Python version information:
```
1.2.1 1.16.1 sys.version_info(major=3, minor=7, micro=1, releaselevel='final', serial=0)
```

### LINKED ISSUE #11474: scipy.stats.skellam.cdf(0) returns 0 for large mu1 = mu2
I need to get `cdf(0)` for skellam distribution with mu1 = mu2. For small mu, `cdf(0)` tends to `0.5` as expected. But for mu more than 5000, it starts to tend to 0, while correct result should tend to 0.5 too. I have checked correct value with [wolfram alpha](https://www.wolframalpha.com/input/?i=+skellam+distribution+6000+6000+cdf%280%29).

#### Reproducing code example:

```
from scipy.stats import skellam

for mu in [1, 10, 100, 1000, 5000, 5050, 5100, 5250, 6000]:
    print(f'mu={mu}, cdf(0)={skellam.cdf(0, mu, mu)}')
```

#### Result is:

```
mu=1, cdf(0)=0.6542541612768373
mu=10, cdf(0)=0.5448901559424116
mu=100, cdf(0)=0.5141135799745626
mu=1000, cdf(0)=0.5044605891382979
mu=5000, cdf(0)=0.5019947363354403
mu=5050, cdf(0)=0.31073482111481376
mu=5100, cdf(0)=0.16106897185096325
mu=5250, cdf(0)=0.006858920449015149
mu=6000, cdf(0)=0.0
```
#### Plots:
plots

#### Scipy/Numpy/Python version information:

1.4.1 1.16.2 sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)

### LINKED ISSUE #12731: stats.ncx2.cdf fails for nc >> x >> 1
`ncx2.cdf(x, df=1, nc=nc)` returns 0.5 when `nc` is much larger than `x` and `x` is much larger than `1`. It should return `0`. For instance: 
```
nc = 10**np.arange(5,10) 
assert np.all(ncx2.cdf(1e4, df=1, nc=nc) == 0.5) 
```

To see that `0` is the correct answer, generate samples of the non-central chisquare distribution from its definition: 

```
from scipy.stats import multivariate_normal, ncx2
import numpy as np 

df = 2
nc = 1e5 
mean = np.ones(df) * np.sqrt(nc/df) 
np.testing.assert_almost_equal( np.linalg.norm(mean)**2, nc) # <-- check that norm(mean)**2 == nc 
xx = multivariate_normal.rvs(size=1000, mean=mean)
X2 = np.linalg.norm(xx,axis=-1)**2 # <-- samples of the non-central chisquare 
threshold = 1e4
assert not np.any(X2 < threshold) 
assert ncx2.cdf(threshold, df=df, nc=nc) > 0.4
```

#### Scipy/Numpy/Python version information:
```
1.5.0 1.18.5 sys.version_info(major=3, minor=7, micro=7, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by caos21**
- Removed `IF` block `tent4`. The removed `IF` block produces the wrong results.
- Expanded search range for `X` and `DF` (as in `CDFLIB`).

#### Reference issue
Closes gh-11474 
Closes gh-12731
Closes gh-8665
Closes gh-9836
Closes gh-2158, although that might have been fixed already

#### What does this implement/fix?
Fixes at least the calculation of the `skellam.cdf` as described in #11474

#### Additional information
The revised code is virtually identical to the `CDFLIB`.

### CODE REVIEW

**Comment by caos21** [comment]
`pnonc` is now bounded by `tent9 = 1e9`. 
This is a new attempt to solve #11474 and #8665 without breaking the tests.

**Comment by tylerjereddy** [comment]
> The revised code is virtually identical to the CDFLIB

Just to clarify: you're saying there's a newer `CDFLIB` implementation somewhere that this code adjustment mimics?

I believe `CDFLIB` is vendored Fortran code, so taking on responsibility for adjusting/maintaining that code may requite a little discussion perhaps?

**Comment by caos21** [comment]
- In my first commit I deleted the `IF` block that does not appears in the `CDFLIB` implementation, and I did 3 minor changes. In that way, the two codes are the same.

- These changes broke an unit test. I restore the conditional but I raised the value for the search in `pnonc`. This fix solves the two issues and pass the test.

I agree that it merits its own discussion. I diffed the two dirs of `CDFLIB` (scipy vs NETLIB) and got some differences.

**Comment by pvanmulbregt** [comment]
The Fortran code in cdfchn.f caps `pnonc` at 10000 and documents the reason as needed to bound the computation.  (This cap also affects gh-9836.)
Computers are faster than they were 20 years ago so raising the limit may be appropriate. What is the change in computation time in the worst case with this PR?

**Comment by caos21** [comment]
You are right, I checked it
```
from scipy.stats import rice
print(rice.cdf(np.arange(10, 150, 10), np.arange(10, 150, 10)))

[0.48002781 0.49002332 0.49335004 0.49501283 0.49601038 0.49667537
 0.49715034 0.49750656 0.49778362 0.49800526 0.49818661 0.49833773
 0.4984656  0.4985752 ]
```
In these tests the elapsed time is almost the same. Maybe we need to
consider a time-consuming test.
I see other errors coming from the difference between implementations and restrictions like this.

It is worth noting that the original implementation does not have these lines:
```
      IF (pnonc.GT.tent4) THEN
          pnonc = tent4
      END IF
```
But it has the search truncated to `tent4`. It seems to be a fix, but it will give the wrong
value for `pnonc` if the condition is met. I think the wrapper for this function should raise an error without changing the logic of the `fortran` code, and this should be reflected in the docs also. The fix proposed only change the value of the search.

**Comment by mdhaber** [comment]
Closes gh-12731
Checking gh-2158, gh-9836, and gh-8665

**Comment by mdhaber** [comment]
Looking very good. I'll put together a speed test and compare with master.

**Comment by mdhaber** [comment]
Starting with 1000 percent points from 0.0001 to 0.9999, I used `chndtrix` to compute the quantiles, and then I used `chndtr` to compute the percent points from the quantiles. I measured the execution time and L2 norm of the error (between given and calculated percentile points) for a range of noncentrality parameter values and degrees of freedom.

The execution time of `chndtr` increases with the square root of the noncentrality parameter (dashed blue line) and is nearly independent of the number of degrees of freedom.

image

The execution time of `chndtrix` also increases with the square root of the noncentrality parameter (dashed blue line) and is nearly independent of the number of degrees of freedom, too.

image

There is an interesting relationship between the "round trip error", the noncentrality parameter, and the number of degrees of freedom. I leave interpretation up to the reader.

image

The total computing time was 138 seconds on the PR branch. I'll try again on master and I'll edit this post based on the comparison.

**Comment by mdhaber** [comment]
Here we go:

image

image

The extra accuracy is not free, but since:

- the execution time only increases when additional computation is required for accuracy (e.g. for low noncentrality parameter, execution time is the same; only when `master` loses accuracy does `chndtr_fix` start taking more time) and
- the execution time only increases with the square root of the noncentrality parameter,

I think it's worth it. 

Ideally, perhaps these limits would be exposed to the user? But I'd suggest merging this PR as-is

**Comment by mdhaber** [comment]
@pvanmulbregt Thanks for finding the relationship between this PR and all those issues! I think I've addressed your question about the change in computation time in the plots above, and I've also added tests demonstrating that this PR fixes those issues. I'd appreciate your review when you have a chance!

@caos21 Thanks very much for this PR! It's surprising how much code these constants affect, and it makes me wonder how many other special function issues we can solve like this.

**Comment by caos21** [comment]
@mdhaber @pvanmulbregt  Thank you, I am sorry I didn't have the time to dig more and doing the performance tests. At that time, I was looking for everything that broke with this implementation, and I don't remember finding anything else. In any case, I will remain more attentive next time :-)

**Comment by WarrenWeckesser** [comment]
@mdhaber's investigation and unit tests demonstrate pretty clearly that this is an improvement to the code.  I've spot checked some values with large `nc` using Wolfram Alpha and `mpsci` and confirmed that this PR is a big improvement. For example,

                         CDF
    x          df   nc   Wolfram & mpsci       This PR              SciPy 1.5.2
    1000001    1    1e6  0.5001994710820217    0.5001994709731743   1.000000000003714
    100000001  1    1e8  0.5000199471139619    0.5000198922867907   1.000000000003714
    100000001  1e3  1e8  0.48010103866913795   0.4801009855849495   1.000000000003714

For `chndtrix`, this computes the equivalent of `chndtrix(p, df, nc)` using `mpmath`:
```
import mpmath
from mpsci.distributions import ncx2

mpmath.mp.dps = 90

df = 100
nc = 1e8
p = 0.46875

x = mpmath.findroot(lambda x: ncx2.cdf(x, df, nc) - p, nc)
print(float(x))
```
It prints `99998530.75750586`, and with this PR, `chndtrix(p, df, nc)` returns `99998530.78259979`, while in SciPy 1.5.2 we get `10083.286532476992` (i.e. garbage).

I'll do a line-by-line persnickety code review, but +1 for the fix and for @mdhaber's unit tests.

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -3412,6 +3412,7 @@ def test_rice_overflow(self):
         assert_(np.isfinite(stats.rice.expect(lambda x: 2, args=(0.74,))))
         assert_(np.isfinite(stats.rice.expect(lambda x: 3, args=(0.74,))))
 
+
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:3415]: Only one blank line before `def`s in a class.
```suggestion
```

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -3553,6 +3554,48 @@ def test_rice_rvs(self):
         assert_equal(rvs(b=3., size=(3, 5)).shape, (3, 5))
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:3556]: Ditto.
```suggestion
```

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -3553,6 +3554,48 @@ def test_rice_rvs(self):
...
+        # options(digits=16)
+        # p = seq(0.1, .9, by = .1)
+        # print(qrice(p, vee = 500, sigma = 4))
+        ppf_exp = [494.8898762347361, 496.6495690858350, 497.9184315188069,
+                   499.0026277378915, 500.0159999146250, 501.0293721352668,
+                   502.1135684981884, 503.3824312270405, 505.1421247157822]
+        assert_allclose(ppf, ppf_exp)
+
+        ppf = scipy.stats.rice.ppf(0.5, np.arange(10, 150, 10))
+         # Generated in R
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:3586]:
```suggestion
        # Generated in R
```

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -4414,6 +4457,39 @@ def test_ncx2_zero_nc_rvs():
     assert_allclose(result, expected, atol=1e-15)
 
 
+def test_ncx2_gh12731():
+    # test that gh-12731 is resolved; previously these were all 0.5
+    nc = 10**np.arange(5,10)
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:4462]:
```suggestion
    nc = 10**np.arange(5, 10)
```

File: scipy/stats/tests/test_distributions.py [hunk 5]
```diff
@@ -4414,6 +4457,39 @@ def test_ncx2_zero_nc_rvs():
...
+    assert_allclose(stats.ncx2.cdf(1e4, df=1, nc=nc), 0)
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:4463]: The true values are so small that the only possible correct answer when using 64 bit floating point is 0, so this should be `assert_equal`.
```suggestion
    assert_equal(stats.ncx2.cdf(1e4, df=1, nc=nc), 0)
```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
I requested a few small changes in-line.  Otherwise, this looks good.

**Comment by caos21** [comment]
Love it. thank you @mdhaber

**Review by WarrenWeckesser — APPROVED** [review]
Thanks @caos21 and @mdhaber, looks good.

**Comment by WarrenWeckesser** [comment]
I'd like to merge the PR this Friday.  Any more comments or questions?  @pv or @person142, what do you think?

**Comment by WarrenWeckesser** [comment]
Thanks @caos21 and @mdhaber, merged.  Always nice to close a bunch of issues with one PR!

**Pull request merged**